File: 프로그래머스/2/12978. 배달/배달.java
import java.util.*;

class Solution {
    public int solution(int N, int[][] road, int K) {
        int answer = 0;
        int[] distanceArr = new int[N+1];
        Arrays.fill(distanceArr, Integer.MAX_VALUE);
        distanceArr[1] = 0;
        
        int[][] graph = new int[N + 1][N + 1];
        for(int i = 0; i <= N; i++){
            Arrays.fill(graph[i], Integer.MAX_VALUE);
        }
        
        for(int[] edge: road){
            int u = edge[0];
            int v = edge[1];
            int dist = edge[2];
            
            graph[u][v] = Math.min(graph[u][v], dist);
            graph[v][u] = Math.min(graph[v][u], dist);
        }
        
        PriorityQueue<int[]> pq = new PriorityQueue<>(new Comparator<int[]>() {
            @Override
            public int compare(int[] o1, int[] o2){
                return o1[0] - o2[0];
            }
        });
        
        // 다익스트라로 1번과 각 정점 최단거리
        int start = 1;
        pq.add(new int[] {distanceArr[1], 1});  // [현재 거리, 현재 정점]
        while(!pq.isEmpty()){
            int curDist = pq.peek()[0];
            int cur = pq.peek()[1];            
            pq.poll();
            
            if(distanceArr[cur] < curDist) continue; // 최적화

            for(int next = 1; next <= N; next++){
                if(graph[cur][next] == Integer.MAX_VALUE) continue;

                int cost = curDist + graph[cur][next];
                if(cost < distanceArr[next]){
                    distanceArr[next] = cost;
                    pq.add(new int[] {cost, next});
                }
            }
        }
        
        for(int i = 1; i <= N; i++){
            if(distanceArr[i] <= K) answer++;
        }
        
        return answer;
    }
}

File: 프로그래머스/2/12978. 배달/배달.py
import heapq
import math

def solution(N, road, K):
    # 거리 설정
    distances = [math.inf for _ in range(N + 1)]
    distances[1] = 0
    
    # 그래프 초기화
    graph = [[math.inf for _ in range(N+1)] for __ in range(N+1)]
    for i in range(1, N+1):
        graph[i][i] = 0
    
    # 그래프 입력, 두 정점 간 경로가 복수이므로 최소 거리 선택
    for u, v, dist in road:
        graph[u][v] = min(graph[u][v], dist)
        graph[v][u] = min(graph[v][u], dist)
    
    # 우선순위 큐, [거리, 정점]
    pq = []
    pq.append([0, 1])
    
    # 반복문 통해 빼기
    while pq:
        dist, vertex = heapq.heappop(pq)
        
        for i in range(1, N+1):
            # 현재 정점은 무시
            if i == vertex:
                continue
                
            dist_between_vertexes, next_vertex = graph[vertex][i], i
            
            # 연결되지 않았다면 무시
            if dist_between_vertexes == math.inf:
                continue
                
            next_dist = dist + dist_between_vertexes
            
            # 현재 거리 + 다음 거리가 현재 출발점부터 거리보다 멀면 무시
            if next_dist > distances[next_vertex]:
                continue

            distances[next_vertex] = next_dist
            heapq.heappush(pq, [next_dist, next_vertex])
    
    return len(list(filter(lambda x: x <= K, distances)))